### PR TITLE
Add `asLiteral` cleaner

### PR DESCRIPTION
**Pull request by samholmes**
This is useful for type narrowing. E.g:

```ts
const asOkResponse = asObject({
  success: asLiteral(true),
  //...
})
const asErrResponse = asObject({
  success: asLiteral(false),
  //...
})

const asResponse = asEither(asOkResponse, asErrResponse)

const response = asResponse(rawResponse)

if (response.success) {
  // Narrowed to asOkResponse
} else {
  // asErrResponse
} 
```

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -197,6 +197,17 @@ export function asMaybe(cleaner, fallback) {
   }
 }
 
+export function asLiteral(literal) {
+  return function asLiteral(raw) {
+    if (raw !== literal) {
+      throw new TypeError(
+        `Expected ${typeof literal} literal '${JSON.stringify(literal)}'`
```
Comment (samholmes) [inline src/index.js]: Change to `Expected ${JSON.stringify(value)}`

File: src/index.d.ts [hunk 2]
```diff
@@ -55,4 +55,8 @@ export declare function asEither<A, B>(
 export declare function asMaybe<T>(cleaner: Cleaner<T>): Cleaner<T | undefined>
 export declare function asMaybe<T>(cleaner: Cleaner<T>, fallback: T): Cleaner<T>
 
+export declare function asLiteral<
```
Comment (samholmes) [inline src/index.d.ts]: Change to `asValue`. Remove `{}` type from union, and add `boolean`. Also, change identifier from `literal` to `value`

**Review by samholmes — COMMENTED** [review]
Aside this feedback, also fold into this PR deprecating `asUndefined` and `asNull` and `asNone`.

**Comment by swansontec** [comment]
In the reference document, we should create a new subsection at the bottom of the Quick Reference section:

```markdown
Deprecated:

- `asMap` - Deprecated alias for `asObject`.
- `asNone` - Use `asOptional(asValue(null))` instead.
- `asNull` - Use `asValue(null)` instead.
- `asUndefined` - Use `asValue(undefined)` instead.
```

**Pull request closed without merging**